cuda/solver/lower_trs_kernels.cu

```diff
@@ -53,10 +58,124 @@ namespace cuda {
+    size_t factor_work_size;
+    void *factor_work_vec;
+};
+static cusp_csrsm2_data cusp_csrsm2_data{};
+
+#elif (defined(CUDA_VERSION) && (CUDA_VERSION < 9020))
+struct cusp_csrsm_data {
+    cusparseSolveAnalysisInfo_t solve_info;
+    cusparseMatDescr_t factor_descr;
+};
+static cusp_csrsm_data cusp_csrsm_data{};
+#endif
+
+
+void clear(std::shared_ptr<const CudaExecutor> exec)
+{
+#if (defined(CUDA_VERSION) && (CUDA_VERSION > 9010))
+    cusparse::destroy(cusp_csrsm2_data.factor_descr);
+    if (cusp_csrsm2_data.solve_info) {
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(
+            cusparseDestroyCsrsm2Info(cusp_csrsm2_data.solve_info));
+    }
+    if (cusp_csrsm2_data.factor_work_vec != nullptr) {
+        exec->free(cusp_csrsm2_data.factor_work_vec);
+    }
+#elif (defined(CUDA_VERSION) && (CUDA_VERSION < 9020))
+    cusparse::destroy(cusp_csrsm_data.factor_descr);
+    GKO_ASSERT_NO_CUSPARSE_ERRORS(
+        cusparseDestroySolveAnalysisInfo(cusp_csrsm_data.solve_info));
+#endif
+}
+
+
 template <typename ValueType, typename IndexType>
 void generate(std::shared_ptr<const CudaExecutor> exec,
               const matrix::Csr<ValueType, IndexType> *matrix,
-              const matrix::Dense<ValueType> *b) GKO_NOT_IMPLEMENTED;
+              const gko::size_type num_rhs)
+{
+    if (cusparse::is_supported<ValueType, IndexType>::value) {
+        auto handle = exec->get_cusparse_handle();
+#if (defined(CUDA_VERSION) && (CUDA_VERSION > 9010))
+        ValueType one = 1.0;
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(
+            cusparseCreateCsrsm2Info(&cusp_csrsm2_data.solve_info));
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(
+            cusparseCreateMatDescr(&cusp_csrsm2_data.factor_descr));
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(cusparseSetMatIndexBase(
+            cusp_csrsm2_data.factor_descr, CUSPARSE_INDEX_BASE_ZERO));
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(cusparseSetMatType(
+            cusp_csrsm2_data.factor_descr, CUSPARSE_MATRIX_TYPE_GENERAL));
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(cusparseSetMatDiagType(
+            cusp_csrsm2_data.factor_descr, CUSPARSE_DIAG_TYPE_NON_UNIT));
+        cusp_csrsm2_data.algorithm = 0;
+        cusp_csrsm2_data.policy = CUSPARSE_SOLVE_POLICY_USE_LEVEL;
+#elif (defined(CUDA_VERSION) && (CUDA_VERSION < 9020))
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(
+            cusparseCreateSolveAnalysisInfo(&cusp_csrsm_data.solve_info));
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(
+            cusparseCreateMatDescr(&cusp_csrsm_data.factor_descr));
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(cusparseSetMatIndexBase(
+            cusp_csrsm_data.factor_descr, CUSPARSE_INDEX_BASE_ZERO));
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(cusparseSetMatType(
+            cusp_csrsm_data.factor_descr, CUSPARSE_MATRIX_TYPE_GENERAL));
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(cusparseSetMatDiagType(
+            cusp_csrsm_data.factor_descr, CUSPARSE_DIAG_TYPE_NON_UNIT));
+#endif
+
+#if (defined(CUDA_VERSION) && (CUDA_VERSION > 9010))
```

Comment:
Why are you opening another `#if` instead of writing the code inside the `#if` above with the exact same condition?
You are not doing anything in between both `#if`s.